File: shade-engine/shade_engine/geo.py
"""지리 좌표 ↔ 로컬 평면(미터) 변환 및 거리/방위 헬퍼.

도시 블록·경로 bbox처럼 좁은 영역에서는 등거리원통도법(equirectangular)
근사로 위경도를 로컬 ENU(동/북) 미터 평면에 충분히 정확히 매핑할 수 있다.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

# WGS84 기준 위도 1도당 거리(m). 경도는 cos(위도)로 보정한다.
_M_PER_DEG_LAT = 111_320.0
_M_PER_DEG_LON_EQUATOR = 111_320.0
EARTH_RADIUS_M = 6_371_000.0


@dataclass(frozen=True)
class LocalProjection:
    """기준점(lat0, lon0) 주변의 평평한 ENU 미터 평면.

    x = 동쪽(+), y = 북쪽(+). 좁은 영역에서만 사용할 것.
    """

    lat0: float
    lon0: float

    @property
    def _m_per_deg_lon(self) -> float:
        return _M_PER_DEG_LON_EQUATOR * math.cos(math.radians(self.lat0))

    def to_xy(self, lat: float, lon: float) -> tuple[float, float]:
        x = (lon - self.lon0) * self._m_per_deg_lon
        y = (lat - self.lat0) * _M_PER_DEG_LAT
        return (x, y)

    def to_latlon(self, x: float, y: float) -> tuple[float, float]:
        lon = self.lon0 + x / self._m_per_deg_lon
        lat = self.lat0 + y / _M_PER_DEG_LAT
        return (lat, lon)


def haversine_m(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """두 위경도 사이의 대권 거리(m)."""
    p1, p2 = math.radians(lat1), math.radians(lat2)
    dphi = math.radians(lat2 - lat1)
    dlmb = math.radians(lon2 - lon1)
    a = math.sin(dphi / 2) ** 2 + math.cos(p1) * math.cos(p2) * math.sin(dlmb / 2) ** 2
    return 2 * EARTH_RADIUS_M * math.asin(min(1.0, math.sqrt(a)))


def bearing_deg(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """지점1→지점2 방위각(도, 북=0 시계방향). [0,360)."""
    p1, p2 = math.radians(lat1), math.radians(lat2)
    dl = math.radians(lon2 - lon1)
    y = math.sin(dl) * math.cos(p2)
    x = math.cos(p1) * math.sin(p2) - math.sin(p1) * math.cos(p2) * math.cos(dl)
    return (math.degrees(math.atan2(y, x)) + 360.0) % 360.0

File: app/app/src/main/java/com/shelter/shade/engine/Geo.kt
package com.shelter.shade.engine

import kotlin.math.*

// WGS84: metres per degree of latitude (constant)
private const val M_PER_DEG_LAT = 111_320.0

// Metres per degree of longitude at the equator (same value, scaled by cos(lat0))
private const val M_PER_DEG_LON_EQUATOR = 111_320.0

/** Mean Earth radius in metres. */
const val EARTH_RADIUS_M = 6_371_000.0

/**
 * Flat-Earth ENU (East / North) projection centred on [lat0], [lon0].
 *
 * x = east (metres, +), y = north (metres, +).
 * Accurate for city-block / walking-path scale areas only.
 *
 * Mirrors Python `geo.LocalProjection`.
 */
class LocalProjection(val lat0: Double, val lon0: Double) {

    /** Metres per degree of longitude at the reference latitude. */
    private val mPerDegLon: Double = M_PER_DEG_LON_EQUATOR * cos(Math.toRadians(lat0))

    /**
     * Convert geographic coordinates to local ENU metres.
     *
     * @return `[x, y]` where x = east metres, y = north metres.
     */
    fun toXy(lat: Double, lon: Double): DoubleArray {
        val x = (lon - lon0) * mPerDegLon
        val y = (lat - lat0) * M_PER_DEG_LAT
        return doubleArrayOf(x, y)
    }

    /**
     * Inverse of [toXy] — convert local ENU metres back to geographic coordinates.
     *
     * @return `[lat, lon]`.
     */
    fun toLatLon(x: Double, y: Double): DoubleArray {
        val lon = lon0 + x / mPerDegLon
        val lat = lat0 + y / M_PER_DEG_LAT
        return doubleArrayOf(lat, lon)
    }
}

/**
 * Great-circle distance between two geographic points using the Haversine formula.
 *
 * @return Distance in metres.
 */
fun haversineM(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
    val p1 = Math.toRadians(lat1)
    val p2 = Math.toRadians(lat2)
    val dPhi = Math.toRadians(lat2 - lat1)
    val dLambda = Math.toRadians(lon2 - lon1)
    val a = sin(dPhi / 2).pow(2) + cos(p1) * cos(p2) * sin(dLambda / 2).pow(2)
    return 2.0 * EARTH_RADIUS_M * asin(min(1.0, sqrt(a)))
}

/**
 * Forward azimuth (bearing) from point 1 to point 2.
 *
 * @return Bearing in degrees, north = 0, clockwise, range [0, 360).
 */
fun bearingDeg(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
    val p1 = Math.toRadians(lat1)
    val p2 = Math.toRadians(lat2)
    val dl = Math.toRadians(lon2 - lon1)
    val y = sin(dl) * cos(p2)
    val x = cos(p1) * sin(p2) - sin(p1) * cos(p2) * cos(dl)
    return (Math.toDegrees(atan2(y, x)) + 360.0) % 360.0
}
